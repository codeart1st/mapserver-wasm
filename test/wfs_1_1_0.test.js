const { loadMapServer, initMapServer, getDocumentFromResponse, printMapServerLog } = require('./utils')

let mapserver

beforeAll(async () => {
  mapserver = await loadMapServer()

  initMapServer(mapserver, {
    mapFileName: 'wfs.map'
  })
})

afterAll(() => {
  //printMapServerLog(mapserver, 'error.log')
})

describe('WFS 1.1.0', () => {
  const globalParams = 'SERVICE=WFS&VERSION=1.1.0'

  test('GetCapabilities request should be valid', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetCapabilities')

    const response = dispatchRequest('GET', searchParams.toString())

    expect(response).toMatchSnapshot()
  })

  test('GetFeature request with unknown feature typename should return a valid ExceptionReport', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetFeature')
    searchParams.set('TYPENAMES', 'unknown')

    const response = dispatchRequest('GET', searchParams.toString())

    expect(response).toMatchSnapshot()
  })

  test('DescribeFeatureType request should be valid', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'DescribeFeatureType')
    searchParams.set('TYPENAMES', 'test')

    const response = dispatchRequest('GET', searchParams.toString())

    expect(response).toMatchSnapshot()
  })

  test('GetFeature request should return a valid feature', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetFeature')
    searchParams.set('TYPENAMES', 'test')
    searchParams.set('MAXFEATURES', '1')

    const response = dispatchRequest('GET', searchParams.toString())

    const document = getDocumentFromResponse(response)

    const result = document.evaluate('//ms:test', document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null)
    expect(result.snapshotLength).toBe(1)

    const node = result.snapshotItem(0)
    expect(node.getAttribute("gml:id")).toBe('test.1')
  })

  test('GetFeature request with Intersects Filter should return a valid feature', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const body = `<?xml version="1.0" encoding="UTF-8"?>
      <wfs:GetFeature xmlns:wfs="http://www.opengis.net/wfs" maxFeatures="10" outputFormat="text/xml; subtype=gml/3.1.1" resultType="results" service="WFS" version="1.1.0">
        <Query xmlns="http://www.opengis.net/wfs" xmlns:ms="http://mapserver.gis.umn.edu/mapserver" srsName="EPSG:4326" typeName="ms:test">
          <ogc:Filter xmlns:ogc="http://www.opengis.net/ogc">
            <ogc:Intersects xmlns:gml="http://www.opengis.net/gml">
              <ogc:PropertyName>ms:msGeometry</ogc:PropertyName>
              <gml:Envelope srsName="EPSG:4326">
                <gml:lowerCorner>-180.0000 -90.0000</gml:lowerCorner>
                <gml:upperCorner>180.0000 90.0000</gml:upperCorner>
              </gml:Envelope>
            </ogc:Intersects>
          </ogc:Filter>
        </Query>
      </wfs:GetFeature>
    `

    const response = dispatchRequest('POST', body)

    const document = getDocumentFromResponse(response)

    const result = document.evaluate('//ms:test', document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null)
    expect(result.snapshotLength).toBe(1)

    const node = result.snapshotItem(0)
    expect(node.getAttribute("gml:id")).toBe('test.1')
  })

  test('GetFeature request with Intersects Filter should return a valid feature with different srs', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const body = `<?xml version="1.0" encoding="UTF-8"?>
      <wfs:GetFeature xmlns:wfs="http://www.opengis.net/wfs" maxFeatures="10" outputFormat="text/xml; subtype=gml/3.1.1" resultType="results" service="WFS" version="1.1.0">
        <Query xmlns="http://www.opengis.net/wfs" xmlns:ms="http://mapserver.gis.umn.edu/mapserver" srsName="EPSG:3857" typeName="ms:test">
          <ogc:Filter xmlns:ogc="http://www.opengis.net/ogc">
            <ogc:Intersects xmlns:gml="http://www.opengis.net/gml">
              <ogc:PropertyName>ms:msGeometry</ogc:PropertyName>
              <gml:Envelope srsName="EPSG:3857">
                <gml:lowerCorner>-20037508 -20037508</gml:lowerCorner>
                <gml:upperCorner>20037508 20037508</gml:upperCorner>
              </gml:Envelope>
            </ogc:Intersects>
          </ogc:Filter>
        </Query>
      </wfs:GetFeature>
    `

    const response = dispatchRequest('POST', body)

    const document = getDocumentFromResponse(response)

    const result = document.evaluate('//ms:test', document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null)
    expect(result.snapshotLength).toBe(1)

    const node = result.snapshotItem(0)
    expect(node.getAttribute("gml:id")).toBe('test.1')

    const result2 = document.evaluate('//gml:pos', document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null)
    expect(result2.snapshotLength).toBe(1)

    const node2 = result2.snapshotItem(0)
    expect(node2.textContent).toBe('222638.981587 6274861.394007')
  })
})