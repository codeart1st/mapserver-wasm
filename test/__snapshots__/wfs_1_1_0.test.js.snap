// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WFS 1.1.0 DescribeFeatureType request should be valid 1`] = `
"Content-Type: text/xml; subtype=gml/3.1.1; charset=UTF-8

<?xml version='1.0' encoding="UTF-8" ?>
<schema
   targetNamespace="http://mapserver.gis.umn.edu/mapserver" 
   xmlns:ms="http://mapserver.gis.umn.edu/mapserver" 
   xmlns:ogc="http://www.opengis.net/ogc"
   xmlns:xsd="http://www.w3.org/2001/XMLSchema"
   xmlns="http://www.w3.org/2001/XMLSchema"
   xmlns:gml="http://www.opengis.net/gml"
   elementFormDefault="qualified" version="0.1" >

  <import namespace="http://www.opengis.net/gml"
          schemaLocation="http://schemas.opengis.net/gml/3.1.1/base/gml.xsd" />

  <element name="test" 
           type="ms:testType" 
           substitutionGroup="gml:_Feature" />

  <complexType name="testType">
    <complexContent>
      <extension base="gml:AbstractFeatureType">
        <sequence>
          <element name="msGeometry" type="gml:GeometryPropertyType" minOccurs="0" maxOccurs="1"/>
          <element name="id" minOccurs="0" type="string"/>
          <element name="name" minOccurs="0" type="string"/>
        </sequence>
      </extension>
    </complexContent>
  </complexType>

</schema>
"
`;

exports[`WFS 1.1.0 GetCapabilities request should be valid 1`] = `
"Content-Type: text/xml; charset=UTF-8

<?xml version="1.0" encoding="UTF-8"?>
<wfs:WFS_Capabilities xmlns:gml="http://www.opengis.net/gml" xmlns:wfs="http://www.opengis.net/wfs" xmlns:ows="http://www.opengis.net/ows" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ogc="http://www.opengis.net/ogc" xmlns="http://www.opengis.net/wfs" version="1.1.0" xsi:schemaLocation="http://www.opengis.net/wfs http://schemas.opengis.net/wfs/1.1.0/wfs.xsd">
  <ows:ServiceIdentification>
    <ows:Title/>
    <!--WARNING: Optional metadata "ows_title" missing for ows:Title-->
    <ows:Abstract/>
    <!--WARNING: Optional metadata "ows_abstract" was missing for ows:Abstract-->
    <!--WARNING: Optional metadata "ows_keywordlist" was missing for ows:KeywordList-->
    <ows:ServiceType codeSpace="OGC">OGC WFS</ows:ServiceType>
    <ows:ServiceTypeVersion>1.1.0</ows:ServiceTypeVersion>
    <ows:Fees/>
    <!--WARNING: Optional metadata "ows_fees" was missing for ows:Fees-->
    <ows:AccessConstraints/>
    <!--WARNING: Optional metadata "ows_accessconstraints" was missing for ows:AccessConstraints-->
  </ows:ServiceIdentification>
  <ows:ServiceProvider>
    <ows:ProviderName/>
    <!--WARNING: Mandatory metadata "ows_contactorganization" was missing for ows:ProviderName-->
    <ows:ProviderSite xlink:type="simple" xlink:href=""/>
    <!--WARNING: Optional metadata "ows_service_onlineresource" was missing for ows:ProviderSite/@xlink:href-->
    <ows:ServiceContact>
      <ows:IndividualName/>
      <!--WARNING: Optional metadata "ows_contactperson" was missing for ows:IndividualName-->
      <ows:PositionName/>
      <!--WARNING: Optional metadata "ows_contactposition" was missing for ows:PositionName-->
      <ows:ContactInfo>
        <ows:Phone>
          <ows:Voice/>
          <!--WARNING: Optional metadata "ows_contactvoicetelephone" was missing for ows:Voice-->
          <ows:Facsimile/>
          <!--WARNING: Optional metadata "ows_contactfacsimiletelephone" was missing for ows:Facsimile-->
        </ows:Phone>
        <ows:Address>
          <ows:DeliveryPoint/>
          <!--WARNING: Optional metadata "ows_address" was missing for ows:DeliveryPoint-->
          <ows:City/>
          <!--WARNING: Optional metadata "ows_city" was missing for ows:City-->
          <ows:AdministrativeArea/>
          <!--WARNING: Optional metadata "ows_stateorprovince" was missing for ows:AdministrativeArea-->
          <ows:PostalCode/>
          <!--WARNING: Optional metadata "ows_postcode" was missing for ows:PostalCode-->
          <ows:Country/>
          <!--WARNING: Optional metadata "ows_country" was missing for ows:Country-->
          <ows:ElectronicMailAddress/>
          <!--WARNING: Optional metadata "ows_contactelectronicmailaddress" was missing for ows:ElectronicMailAddress-->
        </ows:Address>
        <ows:OnlineResource xlink:type="simple" xlink:href=""/>
        <!--WARNING: Optional metadata "ows_service_onlineresource" was missing for ows:OnlineResource/@xlink:href-->
        <ows:HoursOfService/>
        <!--WARNING: Optional metadata "ows_hoursofservice" was missing for ows:HoursOfService-->
        <ows:ContactInstructions/>
        <!--WARNING: Optional metadata "ows_contactinstructions" was missing for ows:ContactInstructions-->
      </ows:ContactInfo>
      <ows:Role/>
      <!--WARNING: Optional metadata "ows_role" was missing for ows:Role-->
    </ows:ServiceContact>
  </ows:ServiceProvider>
  <ows:OperationsMetadata>
    <ows:Operation name="GetCapabilities">
      <ows:DCP>
        <ows:HTTP>
          <ows:Get xlink:type="simple" xlink:href="http://localhost?"/>
          <ows:Post xlink:type="simple" xlink:href="http://localhost?"/>
        </ows:HTTP>
      </ows:DCP>
      <ows:Parameter name="service">
        <ows:Value>WFS</ows:Value>
      </ows:Parameter>
      <ows:Parameter name="AcceptVersions">
        <ows:Value>1.0.0</ows:Value>
        <ows:Value>1.1.0</ows:Value>
      </ows:Parameter>
      <ows:Parameter name="AcceptFormats">
        <ows:Value>text/xml</ows:Value>
      </ows:Parameter>
    </ows:Operation>
    <ows:Operation name="DescribeFeatureType">
      <ows:DCP>
        <ows:HTTP>
          <ows:Get xlink:type="simple" xlink:href="http://localhost?"/>
          <ows:Post xlink:type="simple" xlink:href="http://localhost?"/>
        </ows:HTTP>
      </ows:DCP>
      <ows:Parameter name="outputFormat">
        <ows:Value>XMLSCHEMA</ows:Value>
        <ows:Value>text/xml; subtype=gml/2.1.2</ows:Value>
        <ows:Value>text/xml; subtype=gml/3.1.1</ows:Value>
      </ows:Parameter>
    </ows:Operation>
    <ows:Operation name="GetFeature">
      <ows:DCP>
        <ows:HTTP>
          <ows:Get xlink:type="simple" xlink:href="http://localhost?"/>
          <ows:Post xlink:type="simple" xlink:href="http://localhost?"/>
        </ows:HTTP>
      </ows:DCP>
      <ows:Parameter name="resultType">
        <ows:Value>results</ows:Value>
        <ows:Value>hits</ows:Value>
      </ows:Parameter>
      <ows:Parameter name="outputFormat">
        <ows:Value>text/xml; subtype=gml/3.1.1</ows:Value>
      </ows:Parameter>
    </ows:Operation>
  </ows:OperationsMetadata>
  <FeatureTypeList>
    <Operations>
      <Operation>Query</Operation>
    </Operations>
    <FeatureType>
      <Name>test</Name>
      <Title>test</Title>
      <DefaultSRS>urn:ogc:def:crs:EPSG::4326</DefaultSRS>
      <OtherSRS>urn:ogc:def:crs:EPSG::3857</OtherSRS>
      <OutputFormats>
        <Format>text/xml; subtype=gml/3.1.1</Format>
      </OutputFormats>
      <ows:WGS84BoundingBox dimensions="2">
        <ows:LowerCorner>2.000000 49.000000</ows:LowerCorner>
        <ows:UpperCorner>2.000000 49.000000</ows:UpperCorner>
      </ows:WGS84BoundingBox>
      <MetadataURL format="text/xml" type="TC211">http://localhost?request=GetMetadata&amp;layer=test</MetadataURL>
    </FeatureType>
  </FeatureTypeList>
  <ogc:Filter_Capabilities>
    <ogc:Spatial_Capabilities>
      <ogc:GeometryOperands>
        <ogc:GeometryOperand>gml:Point</ogc:GeometryOperand>
        <ogc:GeometryOperand>gml:LineString</ogc:GeometryOperand>
        <ogc:GeometryOperand>gml:Polygon</ogc:GeometryOperand>
        <ogc:GeometryOperand>gml:Envelope</ogc:GeometryOperand>
      </ogc:GeometryOperands>
      <ogc:SpatialOperators>
        <ogc:SpatialOperator name="Equals"/>
        <ogc:SpatialOperator name="Disjoint"/>
        <ogc:SpatialOperator name="Touches"/>
        <ogc:SpatialOperator name="Within"/>
        <ogc:SpatialOperator name="Overlaps"/>
        <ogc:SpatialOperator name="Crosses"/>
        <ogc:SpatialOperator name="Intersects"/>
        <ogc:SpatialOperator name="Contains"/>
        <ogc:SpatialOperator name="DWithin"/>
        <ogc:SpatialOperator name="Beyond"/>
        <ogc:SpatialOperator name="BBOX"/>
      </ogc:SpatialOperators>
    </ogc:Spatial_Capabilities>
    <ogc:Scalar_Capabilities>
      <ogc:LogicalOperators/>
      <ogc:ComparisonOperators>
        <ogc:ComparisonOperator>LessThan</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>GreaterThan</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>LessThanEqualTo</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>GreaterThanEqualTo</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>EqualTo</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>NotEqualTo</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>Like</ogc:ComparisonOperator>
        <ogc:ComparisonOperator>Between</ogc:ComparisonOperator>
      </ogc:ComparisonOperators>
    </ogc:Scalar_Capabilities>
    <ogc:Id_Capabilities>
      <ogc:EID/>
      <ogc:FID/>
    </ogc:Id_Capabilities>
  </ogc:Filter_Capabilities>
</wfs:WFS_Capabilities>
"
`;

exports[`WFS 1.1.0 GetFeature request with unknown feature typename should return a valid ExceptionReport 1`] = `
"Content-Type: text/xml; charset=UTF-8

<?xml version="1.0" encoding="UTF-8"?>
<ows:ExceptionReport xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ows="http://www.opengis.net/ows" version="1.1.0" language="en-US" xsi:schemaLocation="http://www.opengis.net/ows http://schemas.opengis.net/ows/1.0.0/owsExceptionReport.xsd">
  <ows:Exception exceptionCode="InvalidParameterValue" locator="typename">
    <ows:ExceptionText>msWFSGetFeature(): WFS server error. TYPENAME 'unknown' doesn't exist in this server.  Please check the capabilities and reformulate your request.</ows:ExceptionText>
  </ows:Exception>
</ows:ExceptionReport>
"
`;
